File: backend/app/models/messages.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class BaseMessage(BaseModel):
    type: str
    payload: dict[str, Any] = {}
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
    )
    session_id: str = ""


class ChatMessage(BaseMessage):
    type: Literal["chat.message"] = "chat.message"


class QuestionGroupMessage(BaseMessage):
    type: Literal["chat.question_group"] = "chat.question_group"


class CheckpointMessage(BaseMessage):
    type: Literal["chat.checkpoint"] = "chat.checkpoint"


class StageUpdateMessage(BaseMessage):
    type: Literal["status.stage_update"] = "status.stage_update"


class ProgressMessage(BaseMessage):
    type: Literal["status.progress"] = "status.progress"


class CompleteMessage(BaseMessage):
    type: Literal["status.complete"] = "status.complete"


class ErrorMessage(BaseMessage):
    type: Literal["error.llm_failure", "error.pipeline_failure"] = "error.pipeline_failure"


class ActivityMessage(BaseMessage):
    type: Literal["status.activity"] = "status.activity"


class ControlMessage(BaseMessage):
    type: Literal["control.approve", "control.reject", "control.user_input"]

File: frontend/src/types/messages.ts
export interface BaseMessage {
  type: string
  payload: Record<string, unknown>
  timestamp: string
  session_id: string
}

export interface ChatMessage extends BaseMessage {
  type: "chat.message"
  payload: { text: string }
}

export interface QuestionGroupMessage extends BaseMessage {
  type: "chat.question_group"
  payload: {
    categories: { name: string; confidence: number; questions: string[] }[]
    round: number
    max_rounds: number
  }
}

export interface CheckpointMessage extends BaseMessage {
  type: "chat.checkpoint"
  payload: {
    checkpoint_type: "requirements" | "spec"
    requirements?: Record<string, unknown>
    spec?: Record<string, unknown>
    critique?: Record<string, unknown> | null
    architect_reasoning?: string
  }
}

export interface StageUpdateMessage extends BaseMessage {
  type: "status.stage_update"
  payload: { stage: string; description: string }
}

export interface ProgressMessage extends BaseMessage {
  type: "status.progress"
  payload: { stage: string; percent: number; detail: string }
}

export interface CompleteMessage extends BaseMessage {
  type: "status.complete"
  payload: {
    session_id: string
    framework: string
    download_url: string
    summary: string
  }
}

export interface ErrorMessage extends BaseMessage {
  type: "error.llm_failure" | "error.pipeline_failure"
  payload: { stage: string; message: string; recoverable: boolean }
}

export interface ActivityMessage extends BaseMessage {
  type: "status.activity"
  payload: { agent: string; text: string }
}

export type ServerMessage =
  | ChatMessage
  | QuestionGroupMessage
  | CheckpointMessage
  | StageUpdateMessage
  | ProgressMessage
  | CompleteMessage
  | ErrorMessage
  | ActivityMessage
